libspawner/src/win32/runner.cpp

```diff
@@ -139,26 +139,41 @@ void runner::restore_original_environment(const runner::env_vars_list_t& origina
     }
 }
 
-bool runner::program_stack_exceeds_1gb() {
+void runner::check_executable_segments(bool& excessive_stack, bool& excessive_data) {
     LOADED_IMAGE exe_image;
-    bool stack_exceeds_1gb = false;
-    
     char* exe_path_rw = _strdup(program.c_str());
+
     if (MapAndLoad(exe_path_rw, NULL, &exe_image, FALSE, TRUE)) {
-        stack_exceeds_1gb = 0x40000000 < exe_image.FileHeader->OptionalHeader.SizeOfStackReserve;
+        excessive_stack = 0x40000000 < exe_image.FileHeader->OptionalHeader.SizeOfStackReserve;
+        excessive_data = 0x80000000 < exe_image.FileHeader->OptionalHeader.SizeOfUninitializedData;
         UnMapAndLoad(&exe_image);
+    } else {
+      excessive_stack = false;
+      excessive_data = false;
     }
 
     std::free(exe_path_rw);
-    return stack_exceeds_1gb;
 }
 
 bool runner::try_handle_createproc_error() {
-    DWORD error_code = GetLastError();
+    const DWORD error_code = GetLastError();
+    bool stackseg_excessive, dataseg_excessive;
+
+    // this resets Windows last error, so we restore it
+    check_executable_segments(stackseg_excessive, dataseg_excessive);
+    SetLastError(error_code);
+
+    const bool stackseg_related_error =
+        (error_code == ERROR_NOT_ENOUGH_MEMORY) ||
+        (error_code == ERROR_INVALID_PARAMETER); // Windows XP specific
+
+    const bool dataseg_related_error =
+        (error_code == ERROR_INVALID_PARAMETER) ||
+        (error_code == ERROR_BAD_EXE_FORMAT);
 
     if (
-      error_code == ERROR_NOT_ENOUGH_MEMORY ||
-      (error_code == ERROR_INVALID_PARAMETER && program_stack_exceeds_1gb()) // Windows XP specific
+      (stackseg_related_error && stackseg_excessive) ||
```

Comment:
This is not equivalent, `ERROR_NOT_ENOUGH_MEMORY` should probably still cause ML even without `stackseg_excessive`.